bread/utils.py

```
@@ -0,0 +1,137 @@
+resolves to is a callable, it will be called to get its return
+value, similar to how references to context variables in templates
+work.
+"""
+from inspect import getargspec
+from django.core.exceptions import ValidationError
+from django.db.models import Model
+from django.db.models.fields import FieldDoesNotExist
+
+
+def get_value_or_result(model_instance, attribute_name):
+    attr = getattr(model_instance, attribute_name)
+    if callable(attr):
+        return attr()
+    return attr
+
+
+def get_model_field(model_instance, spec):
+    if model_instance is None:
+        raise ValueError("None passed into get_model_field")
+    if not isinstance(model_instance, Model):
+        raise ValueError("%r should be an instance of a model but it is a %s"
+                         % (model_instance, type(model_instance)))
+
+    name_parts = spec.split('__', 1)
+
+    value = getattr(model_instance, name_parts[0])
+    if callable(value):
+        value = value()
+    if len(name_parts) > 1 and value is not None:
+        return get_model_field(value, name_parts[1])
+    return value
+
+
+def has_required_args(func):
+    """
+    Return True if the function has any required arguments.
+    """
+    spec = getargspec(func)
+    num_args = len(spec.args)
+    # If first arg is 'self', we can ignore one arg
+    if num_args and spec.args[0] == 'self':
+        num_args -= 1
+    # If there are defaults, we can ignore the same number of args
+    if spec.defaults:
+        num_args -= len(spec.defaults)
+    # Do we still have args without defaults?
+    return num_args > 0
+
+
+def validate_fieldspec(model, spec):
+    """
+    Given a model class and a string that refers to a possibly nested
+    field on that model (as used in `get_model_field`).
+
+    Raises a ValidationError with a useful error message if the spec
+    does not appear to be valid.
+
+    Otherwise just returns.
+    """
+    assert issubclass(model, Model)
```

Should you raise a specific Exception type as with other failed sanity checks, or is this "assert" because it would be bread's fault for the error?